pkg/storegateway/indexheader/lazy_binary_reader_test.go

```diff
@@ -383,3 +383,37 @@ func TestLazyBinaryReader_ConcurrentLoadingOfSameIndexReader(t *testing.T) {
 		require.Fail(t, "goroutines did not finish in time")
 	}
 }
+
+func TestLazyBinaryReader_SymbolReaderAndUnload(t *testing.T) {
+	t.Parallel()
+	tmpDir, bkt, blockID := initBucketAndBlocksForTest(t)
+
+	testLazyBinaryReader(t, bkt, tmpDir, blockID, func(t *testing.T, r *LazyBinaryReader, err error) {
+		require.NoError(t, err)
+		require.Nil(t, r.reader)
+		t.Cleanup(func() {
+			require.NoError(t, r.Close())
+		})
+
+		closed := atomic.NewBool(false)
+
+		sr, err := r.SymbolsReader()
+		require.NoError(t, err)
+		go func() {
+			<-time.After(1 * time.Second)
+			require.NoError(t, sr.Close())
+
+			closed.Store(true)
```

Comment:
I think there's potentially a race condition. The `unloadIfIdleSince()` may return right after we call `sr.Close()` above and before this line executes. I would move this line before `sr.Close()`.

Reply:
Done in https://github.com/grafana/mimir/pull/5795/commits/8b12c422a37333d5d2876d34caa3b5039e81bfb4.